#Please edit the following file names and folder names
#Path for reference, databases, softwares and other input files
#Main Working directory - (Doirectory generated by git clone https://github.com/parvathisudha/Targeted-Panel-Analysis.git)
DIR=/N/project/Walker_lab/Targeted_Panel/Code_test/Test2/Targeted-Panel-Analysis
BED_files=$DIR/BED_files
#Reference file for alignment
REF=$DIR/Ref_genome/gatk-bundle/hg38/hg38_chr.fa
#Home directory
Home=/N/u/parkanha/Carbonate
#Path for input files/folders
#Input directory - FASTQ
FASTQ_DIR=$DIR/FASTQ

#For preprocessing
GATK=$Home/gatk-4.1.4.0/gatk
PICARD=$Home/picard-2.10.0_picard.jar
GATK_BUNDLE_DIR=$DIR/Ref_genome/gatk-bundle/hg38
#https://console.cloud.google.com/storage/browser/gatk-software/package-archive/gatk
GATK3=/N/u/parkanha/Carbonate/GenomeAnalysisTK-3.8-1-0-gf15c1c3ef/GenomeAnalysisTK.jar

#For panel analysis
Strelka=$Home/Dataset1_ex/strelka-2.9.2.centos6_x86_64/bin
MANTA=$Home/manta/bin
CNVKIT=$Home/cnvkit/cnvkit
#Source to activate conda-cnvkit (Note this when you download the cnvkit Conda package)
cnvkit=/N/u/parkanha/Carbonate/.conda/envs/cnvkit
fpfilter=$Home/Dataset1_ex
#Ensembl vep files - for variants annotation
VEP=/N/slate/parkanha/WGS/ensembl-vep-release-101/
VEP_FASTA=/N/project/phi_ingest_mm/Brian_Walker/mgp1000_carbonate/mgp1000/references/hg38/Homo_sapiens_assembly38.fasta
VEP_genome=GRCh38
VEP_cache=/N/slate/parkanha/WGS/cache

#Path for input files/folders
#Input directory - FASTQ
FASTQ_DIR=$DIR/FASTQ

BAM_DIR=$DIR/bam
RESULTS_DIR=$DIR/result_files

#Bed files - Muatation,Translocation and combined bed files
#Mutation version
#Change the BED files and version V21 or v22
#If using hg19 change that too
Ver=$BED_files/hg38/v21
Mutation_bed=$Ver/MMmutv21.bed
Translocation_bed=$Ver/MyelomaPanel2Translocationsv2.bed
Mut_Trans_bed=$Ver/Mutv21_Trans.bed

#For Tarpan database
Targed_BED=$Ver/Targeted_regions.bed
Group_BED=$Ver/Groups.bed
Blacklist_BED=$Ver/blacklist.BED
#Database name
DB="Test.db"
#hg19 or hg38
Genome="hg38"
